doc/src/build/programming-with-objects/ch4-object-wrapping.md

```diff
@@ -0,0 +1,195 @@
+
+Let's walk through an example implementation of a trusted swap to demonstrate how to use direct wrapping. Let's say there is an NFT-style `Object` type that has `scarcity` and `style`. `scarcity` determines how scarce the object is (presumably the more scarce the higher its market value); `style` determines the object content/type or how it's rendered. You own some of these objects, and want to trade your objects with others. But to make sure it's a fair trade, you are only willing to trade an object with another one that has identical `scarcity` but different `style` (so that you can collect more styles).
+
+First of all, let's define such an object type:
+```rust
+struct Object has key, store {
+    id: VersionedID,
+    scarcity: u8,
+    style: u8,
+}
+```
+In a real application, we probably would make sure that there is a limited supply of the objects and there is a mechanism to mint them to a list of owners. For simplicity and demonstration purpose, here we will just make it straightforward to create:
+```rust
+public(script) fun create_object(scarcity: u8, style: u8, ctx: &mut TxContext) {
+    let object = Object {
+        id: TxContext::new_id(ctx),
+        scarcity,
+        style,
+    };
+    Transfer::transfer(object, TxContext::sender(ctx))
+}
+```
+Anyone can call `create_object` to create a new object with specified `scarcity` and `style`. The created object will be sent to the signer of the transaction. We will likely also want to be able to transfer the object to others:
+```rust
+public(script) fun transfer_object(object: Object, ctx: &mut TxContext) {
+    Transfer::transfer(object, TxContext::sender(ctx))
+}
+```
+
+Now let's look at how we could enable a swap/trade between your object and others' objects. A straightforward idea is this: define a function that takes two objects from two accounts, and swap their ownership. But this doesn't work in Sui! Recall from Chapter 2 that only object owners can send a transaction to mutate the object. So one person cannot send a transaction that would swap their own object with someone else's object.
+
+In the future, we will likely introduce multi-sig transactions so that two people can sign the same transaction for this type of usecase. However you may not always be able to find someone to swap with right away. A multi-sig transaction won't work in this scenario. Even if you can, you may not want to carry the burden of finding a swap target.
```

Comment:
"Multi-sig" usually refers to requiring multiple keys to send a tx from a single address rather than sending a tx that acts on behalf of multiple addresses. We used to call the idea you are describing ["multi-agent"](https://github.com/diem/dip/blob/main/dips/dip-169.md) transactions.